File: frontend/src/types/change-request.ts
export type ChangeRequestStatus =
  | 'draft'
  | 'submitted'
  | 'under_review'
  | 'approved'
  | 'rejected'
  | 'in_progress'
  | 'completed'
  | 'failed'
  | 'cancelled'

export type ChangeRequestType = 'standard' | 'normal' | 'emergency'

export type ChangeRequestRisk = 'low' | 'medium' | 'high'

export type ChangeRequestPriority = 'low' | 'medium' | 'high' | 'critical'

export interface ChangeRequest {
  id: string
  title: string
  description: string
  status: ChangeRequestStatus
  change_type: ChangeRequestType
  risk_level: ChangeRequestRisk
  priority: ChangeRequestPriority
  requester_id: string
  reviewer_id: string | null
  approver_id: string | null
  implementer_id: string | null
  planned_start_at: string | null
  planned_end_at: string | null
  actual_start_at: string | null
  actual_end_at: string | null
  approved_at: string | null
  rejected_at: string | null
  completed_at: string | null
  rejection_reason: string | null
  rollback_plan: string | null
  created_at: string
  updated_at: string
}

export interface ChangeRequestStatusLog {
  id: string
  change_request_id: string
  from_status: ChangeRequestStatus | null
  to_status: ChangeRequestStatus
  changed_by_id: string
  comment: string | null
  created_at: string
}

export interface ChangeRequestDetail extends ChangeRequest {
  status_logs: ChangeRequestStatusLog[]
}

export interface CreateChangeRequestRequest {
  title: string
  description: string
  change_type?: ChangeRequestType
  risk_level?: ChangeRequestRisk
  priority?: ChangeRequestPriority
  planned_start_at?: string | null
  planned_end_at?: string | null
  rollback_plan?: string | null
  implementer_id?: string | null
}

export interface UpdateChangeRequestRequest {
  title?: string
  description?: string
  change_type?: ChangeRequestType
  risk_level?: ChangeRequestRisk
  priority?: ChangeRequestPriority
  planned_start_at?: string | null
  planned_end_at?: string | null
  rollback_plan?: string | null
  implementer_id?: string | null
}

export interface ApproveChangeRequestRequest {
  comment?: string | null
  reviewer_id?: string | null
}

export interface RejectChangeRequestRequest {
  rejection_reason: string
}

export interface TransitionChangeRequestRequest {
  to_status: ChangeRequestStatus
  comment?: string | null
}

export const STATUS_LABELS: Record<ChangeRequestStatus, string> = {
  draft: '下書き',
  submitted: '提出済み',
  under_review: 'レビュー中',
  approved: '承認済み',
  rejected: '却下',
  in_progress: '実施中',
  completed: '完了',
  failed: '失敗',
  cancelled: 'キャンセル',
}

export const TYPE_LABELS: Record<ChangeRequestType, string> = {
  standard: '標準変更',
  normal: '通常変更',
  emergency: '緊急変更',
}

export const RISK_LABELS: Record<ChangeRequestRisk, string> = {
  low: '低',
  medium: '中',
  high: '高',
}

export const PRIORITY_LABELS: Record<ChangeRequestPriority, string> = {
  low: '低',
  medium: '中',
  high: '高',
  critical: '緊急',
}

export const STATUS_COLORS: Record<ChangeRequestStatus, string> = {
  draft: 'bg-gray-100 text-gray-700',
  submitted: 'bg-blue-100 text-blue-800',
  under_review: 'bg-yellow-100 text-yellow-800',
  approved: 'bg-green-100 text-green-800',
  rejected: 'bg-red-100 text-red-800',
  in_progress: 'bg-purple-100 text-purple-800',
  completed: 'bg-green-200 text-green-900',
  failed: 'bg-red-200 text-red-900',
  cancelled: 'bg-gray-100 text-gray-500',
}

export const RISK_COLORS: Record<ChangeRequestRisk, string> = {
  low: 'bg-green-100 text-green-700',
  medium: 'bg-yellow-100 text-yellow-700',
  high: 'bg-red-100 text-red-700',
}

export const TYPE_COLORS: Record<ChangeRequestType, string> = {
  standard: 'bg-blue-100 text-blue-700',
  normal: 'bg-gray-100 text-gray-700',
  emergency: 'bg-orange-100 text-orange-700',
}

File: backend/app/models/change_request.py
import enum
import uuid
from datetime import datetime
from typing import TYPE_CHECKING, Optional

from sqlalchemy import Boolean, Enum as SAEnum, ForeignKey, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.dialects.postgresql import UUID

from app.core.database import Base

if TYPE_CHECKING:
    from app.models.user import User


class ChangeRequestStatus(str, enum.Enum):
    DRAFT = "draft"
    SUBMITTED = "submitted"
    UNDER_REVIEW = "under_review"
    APPROVED = "approved"
    REJECTED = "rejected"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"


class ChangeRequestType(str, enum.Enum):
    STANDARD = "standard"    # Pre-approved, low-risk routine changes
    NORMAL = "normal"        # Requires CAB review and approval
    EMERGENCY = "emergency"  # Urgent change with expedited approval


class ChangeRequestRisk(str, enum.Enum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"


class ChangeRequestPriority(str, enum.Enum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


class ChangeRequest(Base):
    __tablename__ = "change_requests"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    title: Mapped[str] = mapped_column(String(500), nullable=False)
    description: Mapped[str] = mapped_column(Text, nullable=False)
    status: Mapped[ChangeRequestStatus] = mapped_column(
        SAEnum(ChangeRequestStatus, name="change_request_status"),
        nullable=False,
        default=ChangeRequestStatus.DRAFT,
    )
    change_type: Mapped[ChangeRequestType] = mapped_column(
        SAEnum(ChangeRequestType, name="change_request_type"),
        nullable=False,
        default=ChangeRequestType.NORMAL,
    )
    risk_level: Mapped[ChangeRequestRisk] = mapped_column(
        SAEnum(ChangeRequestRisk, name="change_request_risk"),
        nullable=False,
        default=ChangeRequestRisk.MEDIUM,
    )
    priority: Mapped[ChangeRequestPriority] = mapped_column(
        SAEnum(ChangeRequestPriority, name="change_request_priority"),
        nullable=False,
        default=ChangeRequestPriority.MEDIUM,
    )
    requester_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=False
    )
    reviewer_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=True
    )
    approver_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=True
    )
    implementer_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=True
    )
    planned_start_at: Mapped[Optional[datetime]] = mapped_column(nullable=True)
    planned_end_at: Mapped[Optional[datetime]] = mapped_column(nullable=True)
    actual_start_at: Mapped[Optional[datetime]] = mapped_column(nullable=True)
    actual_end_at: Mapped[Optional[datetime]] = mapped_column(nullable=True)
    approved_at: Mapped[Optional[datetime]] = mapped_column(nullable=True)
    rejected_at: Mapped[Optional[datetime]] = mapped_column(nullable=True)
    completed_at: Mapped[Optional[datetime]] = mapped_column(nullable=True)
    rejection_reason: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    rollback_plan: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        server_default=func.now(), onupdate=func.now(), nullable=False
    )

    requester: Mapped["User"] = relationship("User", foreign_keys=[requester_id])
    reviewer: Mapped[Optional["User"]] = relationship("User", foreign_keys=[reviewer_id])
    approver: Mapped[Optional["User"]] = relationship("User", foreign_keys=[approver_id])
    implementer: Mapped[Optional["User"]] = relationship("User", foreign_keys=[implementer_id])
    status_logs: Mapped[list["ChangeRequestStatusLog"]] = relationship(
        "ChangeRequestStatusLog",
        back_populates="change_request",
        order_by="ChangeRequestStatusLog.created_at",
    )
    cab_votes: Mapped[list["CABVote"]] = relationship(
        "CABVote",
        back_populates="change_request",
        order_by="CABVote.voted_at",
    )
    schedule: Mapped[Optional["ChangeSchedule"]] = relationship(
        "ChangeSchedule",
        back_populates="change_request",
        uselist=False,
    )


class ChangeRequestStatusLog(Base):
    __tablename__ = "change_request_status_logs"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    change_request_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("change_requests.id"), nullable=False, index=True
    )
    from_status: Mapped[Optional[ChangeRequestStatus]] = mapped_column(
        SAEnum(ChangeRequestStatus, name="change_request_status"), nullable=True
    )
    to_status: Mapped[ChangeRequestStatus] = mapped_column(
        SAEnum(ChangeRequestStatus, name="change_request_status"), nullable=False
    )
    changed_by_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=False
    )
    comment: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        server_default=func.now(), nullable=False
    )

    change_request: Mapped["ChangeRequest"] = relationship(
        "ChangeRequest", back_populates="status_logs"
    )
    changed_by: Mapped["User"] = relationship("User")


class CABVoteDecision(str, enum.Enum):
    APPROVE = "approve"
    REJECT = "reject"
    ABSTAIN = "abstain"


class CABVote(Base):
    """CAB member voting record for a change request."""

    __tablename__ = "cab_votes"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    change_request_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("change_requests.id"), nullable=False, index=True
    )
    voter_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=False
    )
    decision: Mapped[CABVoteDecision] = mapped_column(
        SAEnum(CABVoteDecision, name="cab_vote_decision"),
        nullable=False,
    )
    comment: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    voted_at: Mapped[datetime] = mapped_column(
        server_default=func.now(), nullable=False
    )

    change_request: Mapped["ChangeRequest"] = relationship(
        "ChangeRequest", back_populates="cab_votes"
    )
    voter: Mapped["User"] = relationship("User")


class ChangeSchedule(Base):
    """Planned maintenance / deployment window for a change request."""

    __tablename__ = "change_schedules"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    change_request_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("change_requests.id"), nullable=False, unique=True, index=True
    )
    scheduled_start: Mapped[datetime] = mapped_column(nullable=False)
    scheduled_end: Mapped[datetime] = mapped_column(nullable=False)
    environment: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)
    notes: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    confirmed: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    created_at: Mapped[datetime] = mapped_column(
        server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        server_default=func.now(), onupdate=func.now(), nullable=False
    )

    change_request: Mapped["ChangeRequest"] = relationship(
        "ChangeRequest", back_populates="schedule"
    )
